--- Makefile	2010-09-10 15:46:02.000000000 -0700
+++ Makefile	2016-07-17 21:59:37.000000000 -0700
@@ -15,13 +15,12 @@
 SHELL=/bin/sh
 
 # To assist in cross-compiling
-CC=gcc
 AR=ar
 RANLIB=ranlib
 LDFLAGS=
 
 BIGFILES=-D_FILE_OFFSET_BITS=64
-CFLAGS=-Wall -Winline -O2 -g $(BIGFILES)
+CFLAGS=-fpic -fPIC -Wall -Winline -O2 -g $(BIGFILES)
 
 # Where you want it installed when you do 'make install'
 PREFIX=/usr/local
